handler/src/main/java/io/netty/handler/ssl/OpenSslKeyMaterial.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.internal.tcnative.SSL;
+import io.netty.util.AbstractReferenceCounted;
+import io.netty.util.IllegalReferenceCountException;
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.ResourceLeakDetectorFactory;
+import io.netty.util.ResourceLeakTracker;
+
+/**
+ * Holds references to the native key-material that is used by OpenSSL.
+ */
+final class OpenSslKeyMaterial extends AbstractReferenceCounted {
+
+    private static final ResourceLeakDetector<OpenSslKeyMaterial> leakDetector =
+            ResourceLeakDetectorFactory.instance().newResourceLeakDetector(OpenSslKeyMaterial.class);
+    private final ResourceLeakTracker<OpenSslKeyMaterial> leak;
+    private long chain;
+    private long privateKey;
+
+    OpenSslKeyMaterial(long chain, long privateKey) {
+        this.chain = chain;
+        this.privateKey = privateKey;
+        leak = leakDetector.track(this);
```

Comment:
instead of a constructor should we use a static factory method and optionally apply the leak record like in buffer and other cases with some sample rate?

Reply:
The sampling is done by `track` internally so I think its fine... Its also the same that we do in `ReferenceCountedOpenSslEngine`.

Reply:
`ReferenceCountedOpenSslEngine` doesn't use the `leak` record on the "hot path" so it is mostly just a memory concern ... the class hierarchy was also less flexible when `ReferenceCountedOpenSslEngine` was introduced bcz it had to be put "underneath" `OpenSslEngine` (which already existed).

Reply:
I agree... that said I am not convinced it worth it to add  a `LeakAwareOpenSslKeyMaterial` here.... if you are strong on is I can do it of-course.

Reply:
Not strong on it. Always something we can do as a followup since the class is package private.